### PR TITLE
Use lowest float for causal mask

### PR DESCRIPTION
**Pull request by tianleiwu**
### Description

Always set causal mask to the lowest float. 

Note that since huggingface transformers v4.21, gpt2 uses lowest half for FP16, and lowest float for FP32:
https://github.com/huggingface/transformers/blob/66fd3a8d626a32989f4569260db32785c6cbf42a/src/transformers/models/gpt2/modeling_gpt2.py#L199
Assume that most fp16 ONNX models are converted from fp32 models. We decided to use lowest float32 for both half and float model for consistency.

The mask_filter_value only applies to raw attention mask (2D, 3D or 4D). For 1D mask, masked item is 0.0 after softmax so mask filter value is the lowest float for 1D mask.
* For BERT model, when users use 1D mask (required by FMHA) and mask_filter_value is not applicable.
* For GPT-2, when environment variable "ORT_ENABLE_FUSED_CAUSAL_ATTENTION=1", we will convert 2D mask to 1D internally.

### Motivation and Context

https://github.com/microsoft/onnxruntime/issues/12843
https://github.com/microsoft/onnxruntime/issues/14363

### CODE REVIEW

File: onnxruntime/contrib_ops/cuda/bert/attention_softmax.cu [hunk 1]
```diff
@@ -172,7 +165,7 @@ __device__ inline void SoftmaxSmall(const int all_sequence_length,
 
   // threadIdx.x might be larger than all_sequence_length due to alignment to 32x.
   if (threadIdx.x < all_sequence_length) {
-    output[index] = T(thread_data_exp * sum_reverse_block);
+    output[index] = is_valid ? T(thread_data_exp * sum_reverse_block) : T(0.f);
```
Comment (tianleiwu) [inline onnxruntime/contrib_ops/cuda/bert/attention_softmax.cu:168]: Add a check so that the code is more robust:  when all words are masked in a row, this will set masked value to 0.0 instead of NaN (caused by divided by zero of sum).

File: onnxruntime/contrib_ops/rocm/bert/attention_softmax.h [hunk 2]
```diff
@@ -122,21 +122,16 @@ __device__ inline void SoftmaxSmall(const int all_sequence_length,
...
   int end = valid_end;
-  if (is_unidirectional) {
-    int end_unid = all_sequence_length - sequence_length + (blockIdx.x % sequence_length) + 1;
-    if (end_unid <= valid_start) {
-      // In this situation, mask of [0, end_unid) and [valid_start, valid_end) has -10000,
-      //              and [end_unid, valid_start) and [valid_end, all_seq_len) has -20000.
-      // So [0, end_unid) will also have value after softmax.
-      is_valid = threadIdx.x < end_unid;
-    } else {
-      end = min(valid_end, end_unid);
```
Comment (tianleiwu) [inline onnxruntime/contrib_ops/rocm/bert/attention_softmax.h:135]: The old code was for mask filter value of -10000. It does not have parity with huggingface transformers latest code with mask filter value of lowest float.

**Review by wangyems — APPROVED** [review]
